### ISSUE TITLE
gotoDate() is not working perfectly

### ISSUE THREAD

**Issue by kingvhit** [thellmund/Android-Week-View#88]
Related issues #25 
Sorry, I think it does not work perfectly. 
I'm facing this issue so far, but not almost cases.
The problems are when I use the method `goto date` sometimes it's not moving to a specific date, but sometimes work fines.

I think it just works with the day nearest current date like.

Ex: Current date: 19/07/2019 
Expected date: 20/07/2019 | 30/07/2019: Its work
Expected date: 20/01/2017 | 1/06/2019: Not move.

Also, I got new problems with the EventChip with Allday mark, it does not show when the screen become in. 
Tap to item of month view -> Show week view -> go-to date -> show week view -> load Events OK -> Show events OK but Event which is display in all day is not show, if want to show I need scroll a little bit to left or right and scroll back by manually.

I have attached the video, please see it first.
[BugGotoDate.mp4.zip](https://github.com/thellmund/Android-Week-View/files/3409352/BugGotoDate.mp4.zip)

My code goto date :
```
        homeViewModel.currentSolar().observe(viewLifecycleOwner, Observer {
            Timber.d("CurrentSolar: $it")
            handler = Handler().apply {
                postDelayed(object : Runnable {
                    override fun run() {
                        this@WeekHolderFragment.runable = this
                        weekView.goToDate(CalendarUtils.convertPlaintDateToCalendar(it))
                    }

                }, 300)
            }
        })
```

**Comment by kingvhit** [comment]
The change time delay from `300` to `500` maybe work for me, please check the issue of events all day

**Comment by thellmund** [comment]
Can you share where in your Fragment (`onViewCreated()`, `onActivityCreated()`, etc.) and how you initialize `WeekView`?

**Comment by kingvhit** [comment]
I found a reason that I set num of visible before gotoDate
Change to 
```
        // Go to date
        homeViewModel.currentSolar().value?.let {
            weekView.goToDate(CalendarUtils.convertPlaintDateToCalendar(it))
        }
        // Update No of Visible
        updateNoOfVisibility(noOfVisible)
```
And it's work perfectly, thanks.


One more thing, the alldays event does not show when going to a specific date. Need scroll to left or right, please see my Video.

**Comment by thellmund** [comment]
I can’t reproduce that issue in the sample app. Can you please provide me with a [SSCCE](http://sscce.org) where the issue occurs?

**Comment by kingvhit** [comment]
Sorry, Please deffer this issue, I can't reproduce on the sample app. 
It's an effect on my app and didn't find the reason.

Currently, my app can `gotoDate` based `runable` but when paged selected, the Alldays event is not shown until I manually scroll to left or right and scroll back.


One more thing, I'm using `SNAPSHOT` version, after you commit to fix the issue at #89 , I can't sê more than 1 EventChip on the same day.  All of my code here, I'm looking at the issue related by my code or not.

```
class WeekHolderFragment : BaseFragment() {

    companion object {
        const val KEY_WEEKLY_HOLDER = "key_weekly_holder"
        const val DAY_HOLDER = 1
        const val THIRD_DAYS_HOLDER = 3
        const val WEEKLY_HOLDER = 7

        /**
         * Create Fragment and putting number of day
         */
        fun create(dayNum: Int) = WeekHolderFragment().apply {
            arguments = Bundle().apply {
                putInt(KEY_WEEKLY_HOLDER, dayNum)
            }
        }
    }

    private lateinit var homeViewModel: HomeViewModel
    private lateinit var weekViewModel: WeekHolderViewModel
    private lateinit var weekView: WeekView<InstanceEntity>
    private val cachedEvents = HashMap<String, ArrayList<WeekViewDisplayable<InstanceEntity>>?>()
    private var noOfVisible: Int = 7

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val itemView = inflater.inflate(R.layout.fragment_week, container, false)
        weekView = itemView.findViewById(R.id.weekView)
        return itemView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initData()
        initViews()
        initEvents()
        initViewModel()
    }

    fun updateNoOfVisibility(num: Int, isNeedMoveToday: Boolean = false) {
        this.noOfVisible = num
        weekView.numberOfVisibleDays = num
        if (isNeedMoveToday) {
            homeViewModel.currentSolar().value?.let {
                weekView.post { weekView.goToDate(CalendarUtils.convertPlaintDateToCalendar(it)) }
            }
        }
    }

    private fun initData() {
        noOfVisible = arguments?.getInt(KEY_WEEKLY_HOLDER, 7) ?: 7
        homeViewModel = ViewModelProviders.of(requireActivity()).get(HomeViewModel::class.java)
        weekViewModel = ViewModelProviders.of(this).get(WeekHolderViewModel::class.java)
    }

    private fun initViews() {
        // Init Views
        with(weekView) {
            // Set date time interpreter
            dateTimeInterpreter = object : DateTimeInterpreter {
                override fun interpretDate(date: Calendar): String {
                    val dayFormatter = CalendarUtils.getDefaultDateFormat(numberOfVisibleDays)
                    val solarDate = CalendarUtils.convertCalendarToPlainDate(date)
                    val lunarDate = solarDate.transform(VietnameseCalendar::class.java)
                    return dayFormatter.format(date.time).toUpperCase(Locale.getDefault()) +
                            " (${CalendarUtils.edgeInt(lunarDate.dayOfMonth)}/${CalendarUtils.edgeInt(lunarDate.month.number)})"
                }

                override fun interpretTime(hour: Int): String {
                    return CalendarUtils.getDefaultTimeFormat(true).format(Calendar.getInstance().apply {
                        set(Calendar.MONTH, Calendar.JANUARY)
                        set(Calendar.DAY_OF_MONTH, 1)
                        set(Calendar.HOUR_OF_DAY, hour)
                        set(Calendar.MINUTE, 0)
                    }.time)
                }
            }
        }
        // Update No of Visible
        updateNoOfVisibility(noOfVisible)
    }

    private fun initEvents() {
        weekView.setOnMonthChangeListener { startDate, endDate ->
            val startVisibleDate = weekView.firstVisibleDate!!
            val startMonth = (startVisibleDate.clone() as? Calendar)!!.apply {
                set(Calendar.DAY_OF_MONTH, 1)
                add(Calendar.MONTH, -1)
            }
            val endMonth = (startVisibleDate.clone() as? Calendar)!!.apply {
                add(Calendar.MONTH, 1)
                set(Calendar.DAY_OF_MONTH, this.getActualMaximum(Calendar.DAY_OF_MONTH))
            }
            // Check data in cached
            if (cachedEvents[getKey(startMonth, endMonth)] == null) {
                // Clean old data first
                this.cachedEvents.clear()
                weekViewModel.loadInstances(startMonth.timeInMillis, endMonth.timeInMillis)
                    .removeObservers(viewLifecycleOwner)
                // Init empty list first
                this.cachedEvents[getKey(startMonth, endMonth)] = arrayListOf()
                // No-data in cached, fetch data
                weekViewModel.fetchInstances(startMonth.timeInMillis, endMonth.timeInMillis)
                weekViewModel.loadInstances(startMonth.timeInMillis, endMonth.timeInMillis)
                    .observe(viewLifecycleOwner, Observer { results ->
                        this@WeekHolderFragment.cachedEvents[getKey(startMonth, endMonth)]?.clear()
                        this@WeekHolderFragment.cachedEvents[getKey(startMonth, endMonth)]?.addAll(results)
                        weekView.notifyDataSetChanged()
                    })
            }
            cachedEvents[getKey(startMonth, endMonth)]?.filter {
                eventMatches(
                    it.toWeekViewEvent(),
                    startDate.get(Calendar.YEAR),
                    startDate.get(Calendar.MONTH)
                )
            } ?: emptyList()
        }

        weekView.setOnEmptyViewClickListener {
            // startActivity(Intent(requireActivity(), ReminderActivity::class.java))
        }
        weekView.setOnEventClickListener { data, rect ->
            // Timber.d("weekView: OnEventClick: ${data.title}")
        }
        weekView.setScrollListener { date ->
            // Update title and update current solar
            CalendarUtils.convertCalendarToPlainDate(date).let {
                (requireActivity() as? HomeActivity)?.updateToolBarTitle(it)
                // Update current solar
                homeViewModel.currentSolar().postValue(it)
            }
        }
    }

    private fun initViewModel() {
        homeViewModel.currentSolar().observe(viewLifecycleOwner, Observer {
            val currentDate = CalendarUtils.convertPlaintDateToCalendar(it)
            if (weekView.firstVisibleDate == null || !currentDate.isSameDay(weekView.firstVisibleDate!!)) {
                weekView.post { weekView.goToDate(currentDate) }
            }
        })
    }

    private fun getKey(startDate: Calendar, endDate: Calendar) = "${startDate.timeInMillis}-${endDate.timeInMillis}"

    /**
     * Checks if an event falls into a specific year and month.
     * @param event The event to check for.
     * @param year The year.
     * @param month The month.
     * @return True if the event matches the year and month.
     */
    private fun eventMatches(event: WeekViewEvent<*>, year: Int, month: Int): Boolean =
        event.startTime.get(Calendar.YEAR) == year && event.startTime.get(Calendar.MONTH) == month
                || event.endTime.get(Calendar.YEAR) == year && event.endTime.get(Calendar.MONTH) == month
}
```

**Comment by thellmund** [comment]
I can’t reproduce the issue – whenever there are two concurrent events, both are displayed. Please prepare a SSCCE (that is: a small sample app that shows the incorrect behavior) if you want me to take a look.
